### ISSUE TITLE
investigate: training instability on medium dataset (10K samples)

### ISSUE THREAD

**Issue by vzaguskin** [vzaguskin/machinelearne-rs#84]
## Problem

During the WGPU vs CPU comparison benchmark, both backends exhibited training instability (loss exploding to infinity) on the medium dataset (10K samples), while small (1K) and large (20K) datasets trained successfully.

### Observed Behavior

```
Medium (10K) samples:
Epoch 0: loss = 0.041
Epoch 1: loss = 0.020
...
Epoch 10: loss = 0.036
Epoch 11: loss = 0.067
Epoch 12: loss = 0.097
Epoch 13: loss = 0.198
...
Epoch 49: loss = 4140095417.52  <-- Exploded!
```

Both CPU and WGPU backends showed the same divergence pattern, suggesting this is not a backend-specific issue.

### Configuration

- Learning rate: 0.01
- Batch size: 32
- Epochs: 50
- Samples: 10,000 (8,000 train, 2,000 test)
- Features: 8 (standardized)

### Why This Matters

1. The same hyperparameters work fine for 1K and 20K datasets
2. This suggests potential issues with:
   - Learning rate sensitivity at certain dataset sizes
   - Batch normalization or gradient scaling issues
   - Data distribution in the 10K subset

## Investigation Steps

- [ ] Check if this is reproducible with different random seeds
- [ ] Test with lower learning rate (0.001) on 10K dataset
- [ ] Compare data distribution between 1K, 10K, and 20K subsets
- [ ] Check gradient magnitudes during training
- [ ] Try different batch sizes (64, 128)
- [ ] Consider adding gradient clipping

## Potential Fixes

1. Adaptive learning rate based on dataset size
2. Gradient clipping to prevent explosion
3. Better default hyperparameters
4. Add convergence monitoring/early stopping

## Related

- Discovered in PR #82 (WGPU vs CPU comparison benchmark)
- Related example: `lib/examples/wgpu_cpu_comparison.rs`
- May relate to `lib/src/trainer/mod.rs` implementation
